test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TestResultCoordinator.Reports.DirectMethod
+{
+    using System;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.NetworkController;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.TestResults;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using TestResultCoordinator.Reports;
+
+    class DirectMethodReportGenerator : ITestResultReportGenerator
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger(nameof(DirectMethodReportGenerator));
+
+        readonly string trackingId;
+
+        public DirectMethodReportGenerator(
+            string trackingId,
+            string expectedSource,
+            ITestResultCollection<TestOperationResult> expectedTestResults,
+            string actualSource,
+            ITestResultCollection<TestOperationResult> actualTestResults,
+            string resultType,
+            ITestResultComparer<TestOperationResult> testResultComparer,
+            Option<NetworkStatusTimeline> networkStatusTimeline)
+        {
+            this.trackingId = Preconditions.CheckNonWhiteSpace(trackingId, nameof(trackingId));
+            this.ExpectedTestResults = Preconditions.CheckNotNull(expectedTestResults, nameof(expectedTestResults));
+            this.ExpectedSource = Preconditions.CheckNonWhiteSpace(expectedSource, nameof(expectedSource));
+            this.ActualSource = Preconditions.CheckNonWhiteSpace(actualSource, nameof(actualSource));
+            this.ActualTestResults = Preconditions.CheckNotNull(actualTestResults, nameof(actualTestResults));
+            this.TestResultComparer = Preconditions.CheckNotNull(testResultComparer, nameof(testResultComparer));
+            this.ResultType = Preconditions.CheckNonWhiteSpace(resultType, nameof(resultType));
+            this.NetworkStatusTimeline = Preconditions.CheckNotNull(networkStatusTimeline, nameof(networkStatusTimeline));
+        }
+
+        internal string ActualSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> ActualTestResults { get; }
+
+        internal string ExpectedSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> ExpectedTestResults { get; }
+
+        internal string ResultType { get; }
+
+        internal ITestResultComparer<TestOperationResult> TestResultComparer { get; }
+
+        internal Option<NetworkStatusTimeline> NetworkStatusTimeline { get; }
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Start to generate report by {nameof(DirectMethodReportGenerator)} for Sources [{this.ExpectedSource}] and [{this.ActualSource}]");
+
+            ulong networkOnSuccess = 0;
```

Comment:
wrap all these counts into a private class/struct; and using it will help to make code more readable.